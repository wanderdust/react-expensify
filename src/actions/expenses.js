import uuid from "uuid";
import database from "../firebase/firebase";

//ADD EXPENSE
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "",
            note = "",
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref("expenses").push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//REMOVE EXPENSE
export const removeExpense = ({id: expenseId} = {}) => ({
    type: "REMOVE_EXPENSE",
    expenseId
});

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});