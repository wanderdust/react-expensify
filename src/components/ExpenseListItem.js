import React from "react";
import { Link } from "react-router-dom"

export const ExpenseListItem = ({ description, createdAt, amount, id }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>Created at: {createdAt}, amount: {amount}</p>
    </div>
);



export default ExpenseListItem;