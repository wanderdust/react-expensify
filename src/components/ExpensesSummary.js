import React from "react";
import { connect } from "react-redux";
import selectExpenses from "./../selectors/expenses";
import selectExpensesTotal from "./../selectors/expenses-total";
import numeral from "numeral";

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount > 1 ? "expenses" : "expense";
    const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
    return (
        <div>
            <h1>You have {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    const expensesTotal = selectExpensesTotal(expenses);
    return {
        expensesCount: expenses.length,
        expensesTotal
    };
};

export default connect(mapStateToProps)(ExpensesSummary);