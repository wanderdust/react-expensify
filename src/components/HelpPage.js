import React from "react";

const HelpPage = () => (
    <div>
        This is my HELP expense compoenent
    </div>
);

export default HelpPage;