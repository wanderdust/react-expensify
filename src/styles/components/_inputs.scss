.text-input {
    border: 1px solid #cacccd;
    font-size: $font-size-large;
    font-weight: 300;
    height: 46px;
    padding: $s-size;
}

.select {
    @extend .text-input;
    background: white;
}

.textarea {
    @extend .text-input;
    height: 10rem;
}