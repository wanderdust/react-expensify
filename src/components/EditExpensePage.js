import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    };
    handleOnRemove = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push("/");
     };
    render () {
        return (
            <div>
            <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit}
            />  
            <button onClick={this.handleOnRemove}>Remove</button>
        </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (expense) => dispatch(editExpense(expense.id, expense)),
    startRemoveExpense: ({id}) => dispatch(startRemoveExpense({id}))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);