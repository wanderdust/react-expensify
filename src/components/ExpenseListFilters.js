import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "./../actions/filters";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        e.target.value === "date" ? (
            this.props.sortByDate()
        ) : (
            this.props.sortByAmount()
        )
    };
    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange} 
                />

                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId="6ñ5j4kl65ñ4"
                    endDate={this.props.filters.endDate}
                    endDateId="a6s5df46as5df4"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                    displayFormat="DD/MM/YYYY"
                    firstDayOfWeek={1}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);