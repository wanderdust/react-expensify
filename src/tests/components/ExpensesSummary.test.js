import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";

let testProps;

beforeEach(() => {
    testProps = {
        expensesCount: 9,
        expensesTotal: 123456
    }
});

test("should render ExpenseSummary correctly", () => {
    const wrapper = shallow(<ExpensesSummary {...testProps} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpesesSummary text when we have multiple expenses", () => {
    const wrapper = shallow(<ExpensesSummary {...testProps} />);
    const text = wrapper.find("h1").at(0).text();
    expect(text).toBe("You have 9 expenses totalling $1,234.56");
    expect(wrapper).toMatchSnapshot();  
});

test("should render ExpesesSummary text when 1 expense", () => {
    const wrapper = shallow(
        <ExpensesSummary 
            expensesCount={1}
            expensesTotal={1000}
        />
    );
    const text = wrapper.find("h1").at(0).text();
    expect(text).toBe("You have 1 expense totalling $10.00");  
    expect(wrapper).toMatchSnapshot();
});