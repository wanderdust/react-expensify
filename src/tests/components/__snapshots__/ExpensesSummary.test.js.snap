// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly 1`] = `
<div>
  <h1>
    You have 
    9
     
    expenses
     totalling 
    $1,234.56
  </h1>
</div>
`;

exports[`should render ExpesesSummary text when 1 expense 1`] = `
<div>
  <h1>
    You have 
    1
     
    expense
     totalling 
    $10.00
  </h1>
</div>
`;

exports[`should render ExpesesSummary text when we have multiple expenses 1`] = `
<div>
  <h1>
    You have 
    9
     
    expenses
     totalling 
    $1,234.56
  </h1>
</div>
`;
